component/loki/source/heroku/internal/herokutarget/herokutarget.go

```diff
@@ -0,0 +1,199 @@
+	h.config.Server.Log = logging.GoKit(log.With(util_log.Logger, "component", "heroku_drain"))
+
+	srv, err := server.New(h.config.Server)
+	if err != nil {
+		return err
+	}
+
+	h.server = srv
+	h.server.HTTP.Path(h.Endpoint()).Methods("POST").Handler(http.HandlerFunc(h.drain))
+
+	go func() {
+		err := srv.Run()
+		if err != nil {
+			level.Error(h.logger).Log("msg", "heroku drain target shutdown with error", "err", err)
+		}
+	}()
+
+	return nil
+}
+
+func (h *HerokuTarget) drain(w http.ResponseWriter, r *http.Request) {
+	entries := h.handler.Chan()
+	defer r.Body.Close()
+	herokuScanner := herokuEncoding.NewDrainScanner(r.Body)
+	for herokuScanner.Scan() {
+		ts := time.Now()
+		message := herokuScanner.Message()
+		lb := labels.NewBuilder(nil)
+		lb.Set("__heroku_drain_host", message.Hostname)
+		lb.Set("__heroku_drain_app", message.Application)
+		lb.Set("__heroku_drain_proc", message.Process)
+		lb.Set("__heroku_drain_log_id", message.ID)
+
+		if h.config.UseIncomingTimestamp {
+			ts = message.Timestamp
+		}
+
+		// Create __heroku_drain_param_<name> labels from query parameters
+		params := r.URL.Query()
+		for k, v := range params {
+			lb.Set(fmt.Sprintf("__heroku_drain_param_%s", k), strings.Join(v, ","))
+		}
+
+		tenantIDHeaderValue := r.Header.Get("X-Scope-OrgID")
+		if tenantIDHeaderValue != "" {
+			// If present, first inject the tenant ID in, so it can be relabeled if necessary
+			lb.Set(ReservedLabelTenantID, tenantIDHeaderValue)
+		}
+
+		processed := relabel.Process(lb.Labels(nil), h.relabelConfigs...)
+
+		// Start with the set of labels fixed in the configuration
+		filtered := h.Labels().Clone()
+		for _, lbl := range processed {
+			if strings.HasPrefix(lbl.Name, "__") {
+				continue
+			}
+			filtered[model.LabelName(lbl.Name)] = model.LabelValue(lbl.Value)
+		}
+
+		// Then, inject it as the reserved label, so it's used by the remote write client
+		if tenantIDHeaderValue != "" {
+			filtered[ReservedLabelTenantID] = model.LabelValue(tenantIDHeaderValue)
+		}
+
+		entries <- loki.Entry{
+			Labels: filtered,
+			Entry: logproto.Entry{
+				Timestamp: ts,
+				Line:      message.Message,
+			},
+		}
+		h.metrics.herokuEntries.WithLabelValues().Inc()
+	}
+	err := herokuScanner.Err()
+	if err != nil {
+		h.metrics.herokuErrors.WithLabelValues().Inc()
+		level.Warn(h.logger).Log("msg", "failed to read incoming heroku request", "err", err.Error())
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	w.WriteHeader(http.StatusNoContent)
+}
+
+func (h *HerokuTarget) Labels() model.LabelSet {
+	return h.config.Labels
+}
+
+func (h *HerokuTarget) ListenAddress() string {
+	return h.config.Server.HTTPListenAddress
+}
+
+func (h *HerokuTarget) ListenPort() int {
+	return h.config.Server.HTTPListenPort
+}
+
+func (h *HerokuTarget) Endpoint() string {
+	return "/heroku/api/v1/drain"
+}
+
+func (h *HerokuTarget) Ready() bool {
+	return true
```

Comment:
I know we just copied this over from Promtail so don't sweat it, but we could add a TODO to use an atomic boolean and actually report when the server is _really_ ready.